Thir unsafeck fixes

matthewjasper opened the pull request:
- Recognise thread local statics in THIR unsafeck
- Add suggestion for unsafe_op_in_unsafe_fn
- Fix unsafe checking of let expressions

matthewjasper committed:
Recognise thread local statics in THIR unsafeck

matthewjasper committed:
Allow tests with rust-rustfix and revisions

matthewjasper committed:
Add suggestion to THIR unsafe_op_in_unsafe_fn lint

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6656553292/job/18089357044) [(plain)](https://github.com/rust-lang/rust/commit/c40cba5704ebb5126b798f57b2f6c59cf1954d1a/checks/18089357044/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v4' (SHA:b4ffde65f46336ab88eb53be808477a3936bae11)
Download action repository 'actions/upload-artifact@v3' (SHA:a8a3f3ad30e3422c9c7b888a15615d19a852ae32)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
Removing intermediate container 8589e284afdc
 ---> 7f20de8918c4
Step 6/10 : COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
 ---> e7e3d8965e59
Step 7/10 : RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
Collecting binaryornot==0.4.4
  Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
Collecting boolean-py==4.0
  Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180117 sha256=2196c9034bf565528bbb1ee6dad4f753eb813f58822363e6b768f09c73e4d4ff
  Stored in directory: /tmp/pip-ephem-wheel-cache-e9iajy1_/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
    Can't uninstall 'setuptools'. No files were found to uninstall.
Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
Collecting virtualenv
  Downloading virtualenv-20.24.6-py3-none-any.whl (3.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 74.2 MB/s eta 0:00:00
Collecting platformdirs<4,>=3.9.1
  Downloading platformdirs-3.11.0-py3-none-any.whl (17 kB)
Collecting filelock<4,>=3.12.2
  Downloading filelock-3.12.4-py3-none-any.whl (11 kB)
Collecting distlib<1,>=0.3.7
  Downloading distlib-0.3.7-py2.py3-none-any.whl (468 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 136.4 MB/s eta 0:00:00
Installing collected packages: distlib, platformdirs, filelock, virtualenv
Successfully installed distlib-0.3.7 filelock-3.12.4 platformdirs-3.11.0 virtualenv-20.24.6
Removing intermediate container 52f528b16dfd
 ---> 30a24bf05183
Step 8/10 : COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
 ---> 99218680816f
 ---> 99218680816f
Step 9/10 : COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/
 ---> 3bb193f6351f
Step 10/10 : ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
Removing intermediate container 88c3bb7907c4
 ---> 8c8f7b201e5b
Successfully built 8c8f7b201e5b
Successfully tagged rust-ci:latest
Successfully tagged rust-ci:latest
##[endgroup]
Built container sha256:8c8f7b201e5b4d95794f68827dbee01dad68a3cfd3137d6ff0d35fb63542a645
Uploading finished image sha256:8c8f7b201e5b4d95794f68827dbee01dad68a3cfd3137d6ff0d35fb63542a645 to https://ci-caches.rust-lang.org/docker/8849b25aebb63c7041ab10114da59fac9c6c89ff409673e53f6251b7e63c69daeaca7298d30885d05004ab27b231421908523f297222d07a53450f37e4691d72
IMAGE          CREATED          CREATED BY                                      SIZE      COMMENT
8c8f7b201e5b   1 second ago     /bin/sh -c #(nop)  ENV SCRIPT=TIDY_PRINT_DIF…   0B        
99218680816f   2 seconds ago    /bin/sh -c #(nop) COPY file:078ea1d11e7b7cda…   367B      
30a24bf05183   4 seconds ago    |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.9MB    
e7e3d8965e59   10 seconds ago   /bin/sh -c #(nop) COPY file:ac591dd6bc5afa66…   5.33kB    
7f20de8918c4   11 seconds ago   |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.1MB    
---
<missing>      3 weeks ago      /bin/sh -c #(nop)  LABEL org.opencontainers.…   0B        
<missing>      3 weeks ago      /bin/sh -c #(nop)  ARG LAUNCHPAD_BUILD_ARCH     0B        
<missing>      3 weeks ago      /bin/sh -c #(nop)  ARG RELEASE                  0B        

<botocore.awsrequest.AWSRequest object at 0x7f12b924b050>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
https://ci-caches.rust-lang.org/docker/8849b25aebb63c7041ab10114da59fac9c6c89ff409673e53f6251b7e63c69daeaca7298d30885d05004ab27b231421908523f297222d07a53450f37e4691d72
sha256:8c8f7b201e5b4d95794f68827dbee01dad68a3cfd3137d6ff0d35fb63542a645
---
DirectMap4k:      196544 kB
DirectMap2M:     7143424 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished dev [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/6f65201659e54ccd40e3285de3e1b7c989af5cc0/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-6f65201659e54ccd40e3285de3e1b7c989af5cc0-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
    Finished release [optimized] target(s) in 26.00s
##[endgroup]
fmt check
tidy check
tidy error: /checkout/compiler/rustc_mir_build/messages.ftl: message `mir_build_unsafe_not_inherited` appears before `mir_build_unsafe_fn_safe_body`, but is alphabetically later than it
run `./x.py test tidy --bless` to sort the file correctly
tidy error: /checkout/compiler/rustc_mir_build/messages.ftl: message `mir_build_wrap_suggestion` appears before `mir_build_unsafe_op_in_unsafe_fn_borrow_of_layout_constrained_field_requires_unsafe`, but is alphabetically later than it
run `./x.py test tidy --bless` to sort the file correctly
removing old virtual environment
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (23.3.1)
Collecting black==23.3.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 7))
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 47.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 47.4 MB/s eta 0:00:00
Collecting click==8.1.3 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 34))
  Downloading click-8.1.3-py3-none-any.whl (96 kB)
Collecting importlib-metadata==6.7.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 38))
  Downloading importlib_metadata-6.7.0-py3-none-any.whl (22 kB)
  Downloading importlib_metadata-6.7.0-py3-none-any.whl (22 kB)
Collecting mypy-extensions==1.0.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 42))
  Downloading mypy_extensions-1.0.0-py3-none-any.whl (4.7 kB)
Collecting packaging==23.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 46))
  Downloading packaging-23.1-py3-none-any.whl (48 kB)
Collecting pathspec==0.11.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 50))
  Downloading pathspec-0.11.1-py3-none-any.whl (29 kB)
  Downloading pathspec-0.11.1-py3-none-any.whl (29 kB)
Collecting platformdirs==3.6.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 54))
  Downloading platformdirs-3.6.0-py3-none-any.whl (16 kB)
Collecting ruff==0.0.272 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 58))
  Downloading ruff-0.0.272-py3-none-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (5.9 MB)
Collecting tomli==2.0.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 77))
  Downloading tomli-2.0.1-py3-none-any.whl (12 kB)
Collecting typed-ast==1.5.4 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 81))
  Downloading typed_ast-1.5.4-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (877 kB)
  Downloading typed_ast-1.5.4-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (877 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 877.7/877.7 kB 131.1 MB/s eta 0:00:00
Collecting typing-extensions==4.6.3 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 107))
  Downloading typing_extensions-4.6.3-py3-none-any.whl (31 kB)
Collecting zipp==3.15.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 114))
  Downloading zipp-3.15.0-py3-none-any.whl (6.8 kB)
Installing collected packages: zipp, typing-extensions, typed-ast, tomli, ruff, platformdirs, pathspec, packaging, mypy-extensions, click, importlib-metadata, black
Successfully installed black-23.3.0 click-8.1.3 importlib-metadata-6.7.0 mypy-extensions-1.0.0 packaging-23.1 pathspec-0.11.1 platformdirs-3.6.0 ruff-0.0.272 tomli-2.0.1 typed-ast-1.5.4 typing-extensions-4.6.3 zipp-3.15.0
some tidy checks failed
Build completed unsuccessfully in 0:00:53
  local time: Thu Oct 26 15:50:12 UTC 2023
  network time: Thu, 26 Oct 2023 15:50:12 GMT
```

</details>

matthewjasper committed:
Visit patterns in THIR let expressions

This fixes some THIR unsafety checking errors not being emitted for
let expressions in these situations.

cjgillot commented on compiler/rustc_mir_build/src/check_unsafety.rs:
```diff
@@ -543,7 +545,32 @@ impl UnsafeOpKind {
+                UnsafeNotInheritedLintNote {
+                    signature_span,
+                    body_start: tcx.sess.source_map().start_point(body_span).shrink_to_hi(),
+                    body_end: tcx.sess.source_map().end_point(body_span).shrink_to_lo(),
```
Having to keep in sync the logic between here and the `impl AddToDiagnostic` is not great, but I don't really have a better suggestion either. Only keep `body_span` in the struct, output `{ unsafe ` at `body_span.shrink_to_lo()`,  and ` }` at `body_span.shrink_to_hi()`?

cjgillot commented on compiler/rustc_mir_build/src/check_unsafety.rs:
```diff
@@ -543,7 +545,25 @@ impl UnsafeOpKind {
+            suggest_unsafe_block.then(|| {
+                let body_span = tcx.hir().body(parent_owner.body_id().unwrap()).value.span;
+                UnsafeNotInheritedLintNote {
+                    signature_span: tcx
+                        .hir()
+                        .span(tcx.hir().local_def_id_to_hir_id(parent_id.def_id)),
```
```suggestion
                    signature_span: tcx.def_span(parent_id.def_id),
```

cjgillot commented on compiler/rustc_mir_build/src/check_unsafety.rs:
```diff
@@ -543,7 +545,25 @@ impl UnsafeOpKind {
     ) {
+        let parent_id = tcx.hir().get_parent_item(hir_id);
+        let parent_owner = tcx.hir().owner(parent_id);
+        let parent_node: hir::Node<'_> = parent_owner.into();
+        let should_suggest = parent_node.fn_sig().map_or(false, |sig| sig.header.is_unsafe());
```
Nit: add a `fn_sig` method to `hir::OwnerNode`?

cjgillot commented on compiler/rustc_mir_build/src/check_unsafety.rs:
```diff
@@ -35,6 +35,7 @@ struct UnsafetyVisitor<'a, 'tcx> {
     param_env: ParamEnv<'tcx>,
     inside_adt: bool,
     warnings: &'a mut Vec<UnusedUnsafeWarning>,
+    suggest_unsafe_block: bool,
```
Could you add a doc-comment on this field?

cjgillot reviewed (commented):
r=me with 3 nits

matthewjasper commented:
@bors r=cjgillot

bors commented:
:pushpin: Commit 868de8e76bbc6ae1a2b857ff2cdf5dc5cab0eadd has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot 868de8e76bbc6ae1a2b857ff2cdf5dc5cab0eadd -->
<!-- homu: {"type":"Approved","sha":"868de8e76bbc6ae1a2b857ff2cdf5dc5cab0eadd","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 868de8e76bbc6ae1a2b857ff2cdf5dc5cab0eadd with merge a2ffbebeb41a01e779a8de27b4236acb36a7b8af...
<!-- homu: {"type":"BuildStarted","head_sha":"868de8e76bbc6ae1a2b857ff2cdf5dc5cab0eadd","merge_sha":"a2ffbebeb41a01e779a8de27b4236acb36a7b8af"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/6775127964/job/18413703027) [(plain)](https://github.com/rust-lang-ci/rust/commit/a2ffbebeb41a01e779a8de27b4236acb36a7b8af/checks/18413703027/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6775127964/job/18413802730)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/6775127964/job/18413802730","builder_name":"checks-actions"} -->

matthewjasper commented:
@bors retry

bors commented:
:hourglass: Testing commit 868de8e76bbc6ae1a2b857ff2cdf5dc5cab0eadd with merge 61a3eea8043cc1c7a09c2adda884e27ffa8a1172...
<!-- homu: {"type":"BuildStarted","head_sha":"868de8e76bbc6ae1a2b857ff2cdf5dc5cab0eadd","merge_sha":"61a3eea8043cc1c7a09c2adda884e27ffa8a1172"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6783333293/job/18441228260)
Approved by: cjgillot
Pushing 61a3eea8043cc1c7a09c2adda884e27ffa8a1172 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"cjgillot","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6783333293/job/18441228260"},"merge_sha":"61a3eea8043cc1c7a09c2adda884e27ffa8a1172"} -->

rust-timer commented:
Finished benchmarking commit (61a3eea8043cc1c7a09c2adda884e27ffa8a1172): [comparison URL](https://perf.rust-lang.org/compare.html?start=114f1f6838173fdcd016f2ab61f3bf6896bbd85b&end=61a3eea8043cc1c7a09c2adda884e27ffa8a1172&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=114f1f6838173fdcd016f2ab61f3bf6896bbd85b&end=61a3eea8043cc1c7a09c2adda884e27ffa8a1172&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=114f1f6838173fdcd016f2ab61f3bf6896bbd85b&end=61a3eea8043cc1c7a09c2adda884e27ffa8a1172&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -4.5% | [-6.5%, -3.5%] | 3     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=114f1f6838173fdcd016f2ab61f3bf6896bbd85b&end=61a3eea8043cc1c7a09c2adda884e27ffa8a1172&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.5% | [-0.5%, -0.5%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.5% | [-0.5%, -0.5%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=114f1f6838173fdcd016f2ab61f3bf6896bbd85b&end=61a3eea8043cc1c7a09c2adda884e27ffa8a1172&stat=size:linked_artifact)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.0% | [0.0%, 0.0%] | 8     |
| Regressions ❌ <br /> (secondary)  | 0.0% | [0.0%, 0.0%] | 10    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.0% | [0.0%, 0.0%] | 8     |
</details>

**Bootstrap**: 663.784s -> 662.103s (-0.25%)
**Artifact size**: 308.95 MiB -> 308.97 MiB (0.01%)
